### PR TITLE
feat(python): validate Enum categories

### PR DESCRIPTION
**Pull request by alexander-beedie**
Split-out from #13345, and contains only the category-validation checks.
Prevents creation of invalid Enums (eg: with categories of the wrong type or duplicate values).

## Examples

Category type check:
```python
pl.Enum([1.234, 5.678])
# TypeError: Enum categories must be strings; found 1.234
```

Category uniqueness check:
```python
pl.Enum(["misc", "value", "value", "other"])
# ValueError: Enum categories must be unique; found 1 duplicates, eg: 'value'
```

### CODE REVIEW

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:553]: I wonder about this type of pattern: if we're converting into a list anyway, isn't it better to leave it up to the user whether he wants to pay the cost of copying the list? Now we're always copying the list which can get expensive for larger lists. I tested with a list of 1 million 3-character strings and calling `list()` on it takes about `1ms`.

So I think it's better to just accept a `list` here instead of an Iterable, unless we're actually iterating over it.

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:559]: If we're going to present all this information, I would be a bit more wordy here:

```suggestion
                f"Enum categories must be unique"
                "\n\nFound {n_dupes} duplicate{plural}. Most common duplicate: {dupe!r}"
```

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:554]: The happy path should be fastest, so you could do something like:

```
if contains_duplicates(categories):
    ...  # compute the actual duplicates and raise an informative error
```

`contains_duplicates` can be faster than the ordered unique function, either simply `len(set(categories)) == len(categories)` or something [more specialized](https://stackoverflow.com/a/56761643).

And I don't think returning the number of duplicates is really that useful here. Just printing the first duplicate is fine. So you could do something like:

```python
def contains_duplicates(iterable):
    seen = []
    for x in iterable:
        if x in seen:
            return x  # Early exit and return the duplicate
        seen.append(x)
    return False
```

(type signature here is a bit concerning but just to convey the idea)

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:561]: This check doesn't have to depend on the uniqueness check.
```suggestion
        for cat in categories:
```

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:563]: In other places where we check the type, we print the _type_ of the inputs, which I think is nice:

```suggestion
                raise TypeError(f"Enum categories must be strings; found {type(cat).__name__!r}")
```

**Review by stinodego — CHANGES_REQUESTED** [review]
Apparently I am in a nitpicky mood today 😬 

Nice to check for duplicates and stringiness, just some comments about the implementation though.

**Comment by alexander-beedie** [comment]
> Apparently I am in a nitpicky mood today 😬

Didn't happen to make this your New Year's resolution? 
Off to a good start if so! 🤣

File: py-polars/polars/datatypes/classes.py
Comment (alexander-beedie) [reply py-polars/polars/datatypes/classes.py:553]: Hmm; yes, materialisation should only really occur if given a generator (otherwise we'd consume it in the checks). Will tweak accordingly.

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:564]: I think we can simplify further if we build the list during the iteration, then we can accept an `Iterable[str]`:

```suggestion
        categories_list= []
        for cat in categories:
            if cat in categories_list:
                raise ValueError(
                    f"Enum categories must be unique; found duplicate {cat!r}"
                )
            if not isinstance(cat, str):
                raise TypeError(
                    f"Enum categories must be strings; found {cat!r} ({type(cat).__name__})"
                )
            categories_list.append(cat)
```

And then set `self.categories = categories_list`.

I think using a list is just as efficient as using a set, because the faster lookup is offset by the slower insert (not sure about this though).

Comment (alexander-beedie) [reply py-polars/polars/datatypes/classes.py:564]: I'll test ;)

Comment (alexander-beedie) [reply py-polars/polars/datatypes/classes.py:564]: Yup, confirmed - as I suspected, definitely shouldn't do list: `O(n)` vs `O(1)` lookup is not a favourable comparison, and the lookup costs utterly dominate  ;)

**Review by stinodego — APPROVED** [review]
Good to merge if CI passes 👍

**Pull request merged**
